### PR TITLE
Add interval joins to tables

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: python/hail/table.py [hunk 1]
```diff
@@ -1251,8 +1251,7 @@ def types_compatible(left, right):
             left = list(left)
             right = list(right)
             return (types_match(left, right)
-                    or (isinstance(src, MatrixTable)
-                        and len(left) == 1
+                    or (len(left) == 1
                         and len(right) == 1
                         and isinstance(left[0].dtype, tinterval)
                         and left[0].dtype.point_type == right[0].dtype))
```
Comment (danking) [inline python/hail/table.py]: I'm trying to understand this code a bit more so I can competently review this PR.

So we're relaxing the requirement that the source of the expressions are a matrix table. So, previously the following expression was not permitted because the indexer is a table and the index expression is an interval.

```python
t1 = hl.utils.range_table(10)
t2 = hl.utils.range_table(10)
t2 = t2.key_by(span = hl.interval(t2.idx, t2.idx + 5))
t2[t1.idx]
```

but this was OK?

```python
mt = hl.utils.range_matrix_table(10, 10)
t2 = hl.utils.range_table(10)
t2 = t2.key_by(span = hl.interval(t2.idx, t2.idx + 5))
t2[mt.idx]
```

this seems a bit odd to me, but maybe I probably don't understand the implementation strategy.

File: python/hail/table.py [hunk 2]
```diff
@@ -1283,17 +1282,29 @@ def types_compatible(left, right):
             for e in exprs:
                 analyze('Table.index', e, src._row_indices)
 
-            right = self.select(**{uid: self.row})
-            uids = [Env.get_uid() for i in range(len(exprs))]
+            is_key = src.key is not None and len(exprs) == len(src.key) and all(
+                exprs[i] is src._fields[list(src.key)[i]] for i in range(len(exprs)))
```
Comment (danking) [inline python/hail/table.py]: is this more clear as `... for (i, expr) in enumerate(exprs)`?

File: python/hail/table.py [hunk 3]
```diff
@@ -1283,17 +1282,29 @@ def types_compatible(left, right):
...
+            is_interval = (len(self.key) == 1
+                           and isinstance(self.key[0].dtype, hl.tinterval)
+                           and exprs[0].dtype == self.key[0].dtype.point_type)
+
+            if not is_key:
+                uids = [Env.get_uid() for i in range(len(exprs))]
+                all_uids = uids[:]
```
Comment (danking) [inline python/hail/table.py]: why copy `uids`? (that is what this doing right?)

File: python/hail/tests/test_api.py [hunk 4]
```diff
@@ -420,6 +420,15 @@ def test_joins(self):
...
 
+    def test_interval_join(self):
+        left = hl.utils.range_table(50, n_partitions=10)
+        intervals = hl.utils.range_table(4)
+        intervals = intervals.key_by(interval = hl.interval(intervals.idx * 10, intervals.idx * 10 + 5))
+        left = left.annotate(interval_matches = intervals.index(left.key))
+        self.assertTrue(left.all(hl.case()
+                                 .when(left.idx % 10 < 5, left.interval_matches.idx == left.idx // 10)
+                                 .default(hl.is_missing(left.interval_matches))))
+
```
Comment (danking) [inline python/hail/tests/test_api.py]: I think we should also have a test using the bracket syntax.

File: src/main/scala/is/hail/table/Table.scala [hunk 5]
```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
...
+    assert(keySignature.exists(s => s.size == 1 && s.types(0) == intervalType.pointType))
+
+    val leftORVD = rvd match {
+      case ordered: OrderedRVD => ordered
+      case unordered =>
+        OrderedRVD.coerce(
+          new OrderedRVDType(key.get.toArray, key.get.toArray, signature),
+          unordered)
+    }
+
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: extra space

File: src/main/scala/is/hail/table/Table.scala [hunk 6]
```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
...
+    val partBc = hc.sc.broadcast(leftORVD.partitioner)
+    val rightSignature = other.signature
+    val rightKeyFieldIdx = other.keyFieldIdx.get(0)
+    val rightValueFieldIdx = other.valueFieldIdx
+    val partitionKeyedIntervals = other.rvd.boundary.crdd
+      .cflatMap { (ctx, rv) =>
+        val region = ctx.region
+        val r = SafeRow(rightSignature, rv)
+        val ur = new UnsafeRow(rightSignature, rv)
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: I'm not sure I see why we should use the unsafe row? Accessing fields of it shouldn't be substantially different in performance from the `SafeRow` (given that we've already paid the cost to construct the `SafeRow`)

File: src/main/scala/is/hail/table/Table.scala [hunk 7]
```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
...
+        val interval = ur.getAs[Interval](rightKeyFieldIdx)
+        if (interval != null) {
+          val rangeTree = partBc.value.rangeTree
+          val pkOrd = partBc.value.pkType.ordering
+          val wrappedInterval = interval.copy(
+            start = Row(Annotation.copy(intervalType.pointType, interval.start)),
+            end = Row(Annotation.copy(intervalType.pointType, interval.end)))
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: If we used the safe row's interval we wouldn't need to copy the start and end here.

File: src/main/scala/is/hail/table/Table.scala [hunk 8]
```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
...
+    val typToInsert: Type = other.valueSignature
+
+    val (newRVType, ins) = signature.unsafeStructInsert(typToInsert, List(fieldName))
+
+    val partBc = hc.sc.broadcast(leftORVD.partitioner)
+    val rightSignature = other.signature
+    val rightKeyFieldIdx = other.keyFieldIdx.get(0)
+    val rightValueFieldIdx = other.valueFieldIdx
+    val partitionKeyedIntervals = other.rvd.boundary.crdd
+      .cflatMap { (ctx, rv) =>
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: You can get away with just `flatMap` here, you don't seem to use the context or the context's region at all.

File: src/main/scala/is/hail/table/Table.scala [hunk 9]
```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
...
+    }).values
+
+    val localRVRowType = signature
+    val pkIndex = signature.fieldIdx(key.get(0))
+    val newTableType = typ.copy(rowType = newRVType)
+    val newOrderedRVType = new OrderedRVDType(key.get.toArray, key.get.toArray, newRVType)
+    val newRVD = leftORVD.zipPartitionsPreservesPartitioning(
+      newOrderedRVType,
+      zipRDD
+    ) { case (it, intervals) =>
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: no need for case, zipPartitionsPreservesPartitioning takes a binary function

File: src/main/scala/is/hail/table/Table.scala [hunk 10]
```diff
@@ -1138,4 +1139,95 @@ class Table(val hc: HailContext, val tir: TableIR) {
...
+        val value = if (queries.isEmpty)
+          null
+        else
+          queries(0)
+        assert(typToInsert.typeCheck(value))
+
+        rvb.set(rv.region)
+        rvb.start(newRVType)
+
+        ins(rv.region, rv.offset, rvb, () => rvb.addAnnotation(typToInsert, value))
```
Comment (danking) [inline src/main/scala/is/hail/table/Table.scala]: Hmm. This seems like it'd be faster if interval tree's contained encoded bytes rather than `Annotation`. Then we'd just be decoding into the region.

Comment (danking) [reply src/main/scala/is/hail/table/Table.scala]: Hmm, I don't think the decoder can use an `rvb` though, so this might be some non-trivial work to make decoders know how to use rvbs.

File: python/hail/table.py (hunk 1 above)
Comment (tpoterba) [reply python/hail/table.py]: this is correct. This PR makes the first one OK.

Comment (tpoterba) [reply python/hail/table.py]: (also, was only ok if it was `t2[mt.row_idx]`

File: python/hail/table.py (hunk 2 above)
Comment (tpoterba) [reply python/hail/table.py]: definitely

File: python/hail/table.py (hunk 3 above)
Comment (tpoterba) [reply python/hail/table.py]: yes. If we append to `all_uids` somewhere before uids is used, then that would be a big problem. This is a way to keep the code safe.

File: python/hail/tests/test_api.py (hunk 4 above)
Comment (tpoterba) [reply python/hail/tests/test_api.py]: I want to push back on this -- we test the bracket syntax elsewhere, and it just calls `.index()`. It doesn't feel like we need to test bracket syntax everywhere we test `index`. Thoughts?

File: src/main/scala/is/hail/table/Table.scala (hunk 5 above)
Comment (tpoterba) [reply src/main/scala/is/hail/table/Table.scala]: This entire section was copied and modified from MatrixTable.annotateRowsIntervalTable. I didn't want to take the time to abstract out the common pieces from the different ones, especially since Patrick is going to delete all of this when his join stuff is ready.

Comment (tpoterba) [reply src/main/scala/is/hail/table/Table.scala]: I did address your comments though (and apply them to both places)

**Comment by konradjk** [comment]
On this branch (`7b7fac863c8d`),
```
    ht = hl.read_table(ht_path)
    context_mt = hl.read_matrix_table(split_context_mt_path).drop('a_index', 'was_split')
    context_mt = context_mt.annotate_rows(vep=context_mt.vep.drop('colocated_variants'))
    context_mt = hl.filter_intervals(context_mt, [hl.parse_locus_interval('1-22')])
    ht.annotate(**context_mt[ht.key, :]).write(output_ht_path, overwrite)
```
gives:
```
Java stack trace:
java.lang.ArrayIndexOutOfBoundsException: 1
    at is.hail.annotations.Annotation$$anonfun$copy$1.apply(Annotation.scala:46)
    at is.hail.annotations.Annotation$$anonfun$copy$1.apply(Annotation.scala:46)
    at scala.Array$.tabulate(Array.scala:331)
    at is.hail.annotations.Annotation$.copy(Annotation.scala:46)
    at is.hail.rvd.OrderedRVDPartitioner.enlargeToRange(OrderedRVDPartitioner.scala:133)
    at is.hail.rvd.KeyedOrderedRVD.orderedJoin(KeyedOrderedRVD.scala:35)
    at is.hail.rvd.OrderedRVD.orderedJoin(OrderedRVD.scala:119)
    at is.hail.expr.TableJoin.execute(Relational.scala:1828)
    at is.hail.expr.TableMapRows.execute(Relational.scala:1859)
    at is.hail.expr.ir.Interpret$.apply(Interpret.scala:465)
    at is.hail.expr.ir.Interpret$.apply(Interpret.scala:39)
    at is.hail.expr.ir.Interpret$.apply(Interpret.scala:15)
    at is.hail.table.Table.write(Table.scala:899)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
    at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
    at py4j.Gateway.invoke(Gateway.java:280)
    at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
    at py4j.commands.CallCommand.execute(CallCommand.java:79)
    at py4j.GatewayConnection.run(GatewayConnection.java:214)
    at java.lang.Thread.run(Thread.java:748)

Hail version: devel-ce257b532e2c
Error summary: ArrayIndexOutOfBoundsException: 1
```
(this doesn't happen on master)

**Comment by tpoterba** [comment]
This ArrayIndexOutOfBounds is caused by a larger design flaw in Tables and ordering, which was previously masked by the horrible coerce-or-shuffle-every-time we were doing. Will try to nail it down today.

File: python/hail/table.py [hunk 11]
```diff
@@ -1296,17 +1295,29 @@ def types_compatible(left, right):
...
-                return left
+                if not is_key:
+                    left = Table(left._jt.select(ApplyMethod('annotate',
+                                                             left._row._ast,
+                                                             hl.struct(**dict(zip(uids, exprs)))._ast).to_hql(), None, None)).key_by(*uids)
+                if is_interval:
+                    return Table(left._jt.intervalJoin(self._jt, uid))
+                else:
+                    right = self.select(**{uid: self.row})
+                    return Table(left._jt.join(right.distinct()._jt, 'left'))
```
Comment (danking) [inline python/hail/table.py]: patrick mentioned there's a more efficient way to do this, is it easy to do that here?

File: python/hail/table.py [hunk 12]
```diff
@@ -1450,12 +1461,13 @@ def _process_joins(self, *exprs):
         broadcast_f = lambda left, data, jt: Table(left._jt.annotateGlobalJSON(data, jt))
```
Comment (danking) [inline python/hail/table.py]: is that comment about ordering out of date now?

File: python/hail/table.py [hunk 13]
```diff
@@ -1450,12 +1461,13 @@ def _process_joins(self, *exprs):
...
         left, cleanup = process_joins(self, exprs, broadcast_f)
 
-        if left is not self:
+        if left is not self and (original_key is None and left.key is not None
+                                 or original_key is not None and (left.key is None or list(left.key) != original_key)):
```
Comment (danking) [inline python/hail/table.py]: this all seems like a very complicated way of saying `original_key != left.key` but somehow `None` and `list` screw that up? Is there a way to make this less complicated? Or maybe if we had a `is_same_key` function? What exactly is a key? It's not `None or list(type)`, it seems like `None or ???` but I don't know what the ??? is.

File: python/hail/table.py [hunk 14]
```diff
@@ -1450,12 +1461,13 @@ def _process_joins(self, *exprs):
...
             if original_key is not None:
                 if original_key != list(left.key):
                     left = left.key_by(*original_key)
             else:
-                left = left.key_by(None)
+                left = left.key_by()
```
Comment (danking) [inline python/hail/table.py]: kind of disappointing we can't just do `key_by(original_key)`

File: python/hail/table.py [hunk 15]
```diff
@@ -2328,7 +2340,7 @@ def row_value(self) -> 'StructExpression':
         :class:`.StructExpression`
             Struct of all row fields.
         """
-        return self._row.drop(*self.key.keys()) if self.key else self._row
+        return self._row.drop(*self.key.keys()) if self.key is not None else self._row
```
Comment (danking) [inline python/hail/table.py]: the key possibly being None is really a drag.

File: src/main/scala/is/hail/rvd/OrderedRVD.scala [hunk 16]
```diff
@@ -99,7 +99,6 @@ class OrderedRVD(
 
     require(ordType.rowType == typ.rowType)
     require(ordType.kType isPrefixOf typ.kType)
-    require(newPartitioner.pkType isIsomorphicTo ordType.pkType)
```
Comment (danking) [inline src/main/scala/is/hail/rvd/OrderedRVD.scala]: Why is this no longer required?

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala [hunk 17]
```diff
@@ -18,8 +18,6 @@ class OrderedRVDPartitioner(
   val pkIntervalType = TInterval(pkType)
   val rangeBoundsType = TArray(pkIntervalType)
 
-  assert(rangeBoundsType.typeCheck(rangeBounds))
```
Comment (danking) [inline src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: This seems like an OK thing to check?

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala [hunk 18]
```diff
@@ -130,8 +128,8 @@ class OrderedRVDPartitioner(
       return this
     if (range.isEmpty)
       return copy(rangeBounds = FastIndexedSeq(newRange.get))
-    val newStart = pkType.ordering.min(range.get.start, Annotation.copy(pkType, newRange.get.start))
-    val newEnd = pkType.ordering.max(range.get.end, Annotation.copy(pkType, newRange.get.end))
```
Comment (danking) [inline src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: So two things stick out to me. `enlargeToRange` is only every called on a partitioner's `range`, so this relaxation is safe as long as partitioner's range is always safe, which boils down to `rangeBounds` always being safe.

This is kind of an uncomfortable amount of assumptions and digging that we have to ensure is still correct with every new PR. I'm not suggesting we fix that in _this_ PR. Can you share a bit of the motivation for removing this particular copy?

File: src/test/scala/is/hail/vds/JoinSuite.scala [hunk 19]
```diff
@@ -2,6 +2,7 @@ package is.hail.vds
 
 import is.hail.SparkSuite
 import is.hail.annotations._
+import is.hail.expr.MatrixRowsTable
```
Comment (danking) [inline src/test/scala/is/hail/vds/JoinSuite.scala]: I think this an unnecessary change.

File: python/hail/table.py (hunk 13 above)
Comment (tpoterba) [reply python/hail/table.py]: it's None or a StructExpression. I can simplify this though.

File: python/hail/table.py (hunk 12 above)
Comment (tpoterba) [reply python/hail/table.py]: ya, it's in the wrong place.

File: python/hail/table.py (hunk 14 above)
Comment (tpoterba) [reply python/hail/table.py]: I concluded this is a bug keying at the end -- I've moved it into the join function.

File: python/hail/tests/test_api.py (hunk 4 above)
Comment (danking) [reply python/hail/tests/test_api.py]: As discussed out of line, __getitem__ is simple enough in the case that you're just passing an expression or list of expressions that this is OK with me.

File: src/main/scala/is/hail/rvd/OrderedRVD.scala (hunk 16 above)
Comment (tpoterba) [reply src/main/scala/is/hail/rvd/OrderedRVD.scala]: Patrick can explain it better, but the idea is that a something partitioned by {locus, alleles} can be joined with something partitioned by {locus} (as {locus, alleles} is a subtype of {locus}). My table key changes in Python caused problems to pop up that were previously being masked by the horribly inefficient join strategy (re-key every time). This is the solution to that.

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala (hunk 17 above)
Comment (tpoterba) [reply src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: I think it's typeCheck that's broken in this case, because `interval<struct{a}>` should let `interval<struct{a, b}>` through. Patrick has been thinking about this more.

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala (hunk 18 above)
Comment (tpoterba) [reply src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: also related to subtyping above -- `newRange.get.start` could be a subtype of `pkType` and we're not handling that.

File: src/main/scala/is/hail/rvd/OrderedRVD.scala (hunk 16 above)
Comment (danking) [reply src/main/scala/is/hail/rvd/OrderedRVD.scala]: @patrick-schultz is there a different requirement we should check here?

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala (hunk 17 above)
Comment (danking) [reply src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: Hmm. Maybe not typeCheck itself, but perhaps there should be a different way to check this relationship exists, and, if necessary, what code is needed to bridge the physical implementation and the actual implementation.

File: src/main/scala/is/hail/rvd/OrderedRVD.scala (hunk 16 above)
Comment (patrick-schultz) [reply src/main/scala/is/hail/rvd/OrderedRVD.scala]: I've been trying to build the join infrastructure so that there is no requirement on the partition keys whatsoever (and I think eventually they will go away). I think https://github.com/hail-is/hail/pull/3442/files#diff-cc90b5eede7c86a868dd954853660e37L40 enabled this to be removed, and I just missed it.

File: src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala (hunk 17 above)
Comment (patrick-schultz) [reply src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: I'm not sure we want to change the behavior of typeCheck, but there is a more relaxed check we could make here.

Comment (danking) [reply src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala]: Ok. what we need here is a `TBaseStruct.isComparableAt` (naming is hard) that we can use something like:
```scala
  assert(rangeBounds.forall { case Interval(x,y,_,_) =>
    pkType.isComparableAt(x) && pkType.isComparableAt(y)
  }
```
It should mirror the implementation of `ExtendedOrdering.rowOrdering`, checking that at least the first `pkType.fields.length` worth of columns type check against the corresponding fields.

**Comment by tpoterba** [comment]
@patrick-schultz can I offload this to you to address Dan's comment and get it in? You should be able to push to this branch. I need to spend the time I have left before vacation working on some data relocation that's taken longer than I anticipated, and I don't think this PR can wait two weeks (Konrad is already using it!)

**Comment by patrick-schultz** [comment]
@tpoterba No problem.
